Add SCC support for blocking pods from using EmptyDir volumes.

dgoodwin opened the pull request:
Builds off the example of AllowHostDirVolumePlugin. Change is being implemented to satisfy some online/multi-tenant use cases, and roughly matches a planned feature for pod security policy: https://github.com/pweil-/kubernetes/blob/security-policy/docs/proposals/security-context-constraints.md

There is however an outstanding problem I could use some guidance on, today use of empty dir volumes is implicitly allowed, but this SCC parameter is a boolean to allow it, which will zero to false effectively breaking every SCC out in the wild. I am unsure how best to roll this out.

Ideas:

(1) reconcile-sccs command

(2) Inverting it to DenyEmptyDirVolumePlugin. There's nothing else in the SCC that would match this 'deny' convention, so it feels inconsistent, but it might be more technically correct for what it actually needs to do.

(3) Alternatively could we make it a string so we can have a legitimate third state? Empty string = unset, "true" and "false"? 

We don't really need a third state so in some ways I'd kind of lean towards (2) as being the best option, but please let me know what you think or if there are other ideas.

liggitt commented:
current SCCs allow empty dir, so we have to preserve that behavior

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/types.go:
```diff
@@ -2601,6 +2601,8 @@ type SecurityContextConstraints struct {
 	// AllowHostDirVolumePlugin determines if the policy allow containers to use the HostDir volume plugin
 	AllowHostDirVolumePlugin bool `json:"allowHostDirVolumePlugin" description:"allow the use of the host dir volume plugin"`
+	// AllowEmptyDirVolumePlugin determines if the policy allow containers to use the EmptyDir volume plugin
+	AllowEmptyDirVolumePlugin bool `json:"allowEmptyDirVolumePlugin" description:"allow the use of the empty dir volume plugin"`
```
if you need to distinguish between unset, true, and false, you can make the serialized field a *bool, and use defaulting to preserve prior behavior

ironcladlou replied:
I'm not we need to distinguish unset with a pointer type. I think we can use a defaulting function to set `true`.

ironcladlou replied:
Whoops, nevermind. We need a pointer AND a defaulter.

ironcladlou replied:
Existing SCC defaults: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/defaults.go#L276

dgoodwin commented:
Thanks for the tips @liggitt & @ironcladlou 

Branch & commit message updated, now using a primitive bool on the API SCC but *bool on the versioned SCC types. 

Registered a default for the type as well to ensure we make this true whenever possible. The bootstrappolicy still needs to set it explicitly however as we're just creating the objects directly.

The k8s "*generated*".go files are manually updated for the new struct field. 

Removing WIP as I think it's ready for full review.

ironcladlou commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go:
```diff
@@ -251,6 +251,14 @@ func (s *simpleProvider) ValidateContainerSecurityContext(pod *api.Pod, containe
+	if !s.scc.AllowEmptyDirVolumePlugin {
+		for _, v := range pod.Spec.Volumes {
+			if v.EmptyDir != nil {
+				allErrs = append(allErrs, fielderrors.NewFieldInvalid("VolumeMounts", v.Name, "EmptyDir Volumes are not allowed to be used"))
```
I don't think this is the canonical format for a field error. Check out https://github.com/dgoodwin/origin/blob/block-emptydir-scc/Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go#L264 for reference. I think you should address it as something like `volumes[index]` (and correctly prefixed).

dgoodwin replied:
Error creating: Pod "testrc-" is forbidden: unable to validate against any security context constraint: [provider restricted: .spec.containers[0].securityContext.VolumeMounts: invalid value 'container-data', Details: EmptyDir Volumes are not allowed to be used]

I changed it to this now:

Error creating: Pod "testrc-" is forbidden: unable to validate against any security context constraint: [provider restricted: .spec.containers[0].securityContext.volumes[0]: invalid value 'emptyDir', Details: EmptyDir volumes are not allowed to be used]

I also adjusted the HostPath error just above where I got the bad example originally.

ironcladlou commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -222,6 +228,7 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
 
 	// add default access
 	for i, constraint := range constraints {
+
```
Unnecessary whitespace

dgoodwin commented:
Updated for review, separate commit for now to help reviewers, will squash before merge.

ironcladlou commented:
LGTM, but probably need a final review from @liggitt or @pweil-.

liggitt commented on hack/update-generated-conversions.sh:
```diff
@@ -29,7 +29,7 @@ EOF
 	cat >> $TMPFILE <<EOF
 // AUTO-GENERATED FUNCTIONS END HERE
 EOF
-	
```
undo change

liggitt replied:
nm, thought that was under Godeps

liggitt replied:
undo whitespace change

liggitt commented on docs/proposals/security-context-constraints.md:
```diff
@@ -90,6 +90,8 @@ type SecurityContextConstraints struct {
 	// AllowHostDirVolumePlugin determines if the policy allow containers to use the HostDir volume plugin
 	AllowHostDirVolumePlugin bool `json:"allowHostDirVolumePlugin,omitempty" description:"allow the use of the host dir volume plugin"`
+	// AllowEmptyDirVolumePlugin determines if the policy allow containers to use the EmptyDir volume plugin
+	AllowEmptyDirVolumePlugin bool `json:"allowEmptyDirVolumePlugin,omitempty" description:"allow the use of the empty dir volume plugin"`
```
undo change

liggitt commented:
split into three commits:
1. all v1beta3 changes in a commit titled "UPSTREAM: <carry>: v1beta3 scc"
2. all other Godep changes in a commit titled "UPSTREAM: <carry>: scc"
3. all other changes in a regular commit

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/types.go:
```diff
@@ -2601,6 +2601,8 @@ type SecurityContextConstraints struct {
 	// AllowHostDirVolumePlugin determines if the policy allow containers to use the HostDir volume plugin
 	AllowHostDirVolumePlugin bool `json:"allowHostDirVolumePlugin" description:"allow the use of the host dir volume plugin"`
+	// AllowEmptyDirVolumePlugin determines if the policy allow containers to use the EmptyDir volume plugin
+	AllowEmptyDirVolumePlugin *bool `json:"allowEmptyDirVolumePlugin" description:"allow the use of the empty dir volume plugin"`
```
the incongruence makes me very sad, but I don't see a better way to keep the current behavior

pweil- replied:
This makes me wonder if we should change the upstream proposal to be a black list of plugins rather than a whitelist.

dgoodwin commented:
Updated into three separate commits, rolled back whitespace change to script I didn't end up modifying. 

@liggitt I saw in email you were thinking about removing the change to the proposal .md file, but looks like you removed it. If github ate that somehow just say the word and I'll pull it out as well.

dgoodwin commented:
Looks like something doesn't like those commit msg formats, would "UPSTREAM: <carry>: v1beta3 scc" be correct here? Or is this a "<none>"? Looks like most SCC commits are marked <carry>.

liggitt commented:
I'd still like @pweil-'s feedback, and if possible, to hold this until the rebase lands

pweil- commented:
[test]

dgoodwin commented:
Brought the volume scanning into one loop guarted on both conditions, and my apologies that's a legit test failure, I somehow slipped breakage in while refactoring during review. Should be healthy now.

dgoodwin commented:
Updated with a master rebase and some things @dobbymoodge caught.

- renamed boolPtr helper to newBool to better match some other similar functions.
- added a missed assertion in the v1beta3 defaults_test.go.
- more closely match the spirit of some existing tests in both defaults_test.go files by explicitly setting an allowEmptyDirVolumePlugin value.

Still waiting on kube rebase for merge.

dobbymoodge commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1beta3/defaults_test.go:
```diff
@@ -470,23 +470,32 @@ func areSecurityContextAndContainerEqual(c *versioned.Container) (bool, []string
-			expectedSupGroup: versioned.SupplementalGroupsStrategyMustRunAs,
+			expectedFSGroup:       versioned.FSGroupStrategyMustRunAs,
+			expectedSupGroup:      versioned.SupplementalGroupsStrategyMustRunAs,
+			expectedAllowEmptyDir: true,
```
Shouldn't this be `expectedAllowEmptyDir: false,`, since default is `true`?

dgoodwin replied:
This would then be identical to "preserve AllowEmptyDirVolumePlugin set to false" at the end, the only case missing was set to true and check true. Probably shouldn't be messing around with this test at all, it's named generically but I don't think it's actually meant for every property possible. Instead I should probably just add another test at the end for set true check true. Set nothing and expect true would be covered by all pre-existing tests.

dgoodwin commented:
Rebased on top of new k8s, test added per comments above.

Should be GTG on my side @liggitt .

dgoodwin commented:
re[test] please.

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go:
```diff
@@ -243,10 +243,17 @@ func (s *simpleProvider) ValidateContainerSecurityContext(pod *api.Pod, containe
-				allErrs = append(allErrs, field.Invalid(fldPath.Child("volumeMounts"), v.Name, "Host Volumes are not allowed to be used"))
+			if !s.scc.AllowHostDirVolumePlugin && v.HostPath != nil {
+				allErrs = append(allErrs, field.Invalid(
+					fldPath.Child("volumes"), "hostPath",
```
...Child("volumes").Index(i)

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go:
```diff
@@ -243,10 +243,17 @@ func (s *simpleProvider) ValidateContainerSecurityContext(pod *api.Pod, containe
+			}
+			if !s.scc.AllowEmptyDirVolumePlugin && v.EmptyDir != nil {
+				allErrs = append(allErrs, field.Invalid(
+					fldPath.Child("volumes"), "emptyDir",
```
same here

dgoodwin commented:
@liggitt updated, thx.

dgoodwin commented:
Re[test] please.

dgoodwin commented:
Robot, re[test] it again please. 

![687474703a2f2f6d656469612e67697068792e636f6d2f6d656469612f416e764f31784b616131386d512f67697068792e676966](https://cloud.githubusercontent.com/assets/51265/12684279/6efc1646-c694-11e5-9769-8a7c91359396.gif)

liggitt commented:
seems ok to me, @pweil- has final LGTM
@deads2k, I assume we're not at the point of holding this out for the rebase yet?

pweil- commented:
minor updates and LGTM

dgoodwin commented:
@pweil- @liggitt Updated JSON description to note the default is true, modified "shouldn't default" test as requests and deleted the test this obsoleted, added success case test.

dgoodwin commented:
Not sure what's going on, the verify swagger script is rejecting the output of the update swagger script. Need to run for the day, will get this sorted tomorrow.

pweil- commented:
> Not sure what's going on, the verify swagger script is rejecting the output of the update swagger script. Need to run for the day, will get this sorted tomorrow.

you may need to rebase and regenerate

deads2k commented:
> @deads2k, I assume we're not at the point of holding this out for the rebase yet?

No, the queue is still open.  This could take a while.

dgoodwin commented:
My local swagger diff problem was caused by a vim .swp file floating around in there.

However I am really not sure why it's still failing here.

$ sudo hack/verify-generated-swagger-spec.sh
===== Verifying API Swagger Spec =====
Generating a fresh spec...
Diffing current spec against freshly generated spec...
SUCCESS: Swagger spec up to date.

Travis is claiming a diff on a few fields unrelated to this PR... trying to sort it out.

dgoodwin commented:
Alright that looks like a test infra flake. Re-[test] please.

dgoodwin commented:
Re[test] please.

dgoodwin commented:
[test]

pweil- commented:
[merge]

deads2k commented:
We're down to unit tests now.  I'd like a hold on this until after the rebase (hopefully this week/early next week).  Particularly since we're changing up how UPSTREAM patches are handled for rebases.

dgoodwin commented:
Reapplied on top of the rebase but I don't think this is ready yet until master settles down, I'm seeing two issues that sound master related, one the default is coming out "false" suddenly, and two any attempt to edit an scc is met with a "apiVersion should not be changed" error. Will keep trying as master settles down.

deads2k commented:
> Reapplied on top of the rebase but I don't think this is ready yet until master settles down, I'm seeing two issues that sound master related, one the default is coming out "false" suddenly, and two any attempt to edit an scc is met with a "apiVersion should not be changed" error. Will keep trying as master settles down.

I'm not aware of either of those issues.  Can you write them up with details so we can look at them?

dgoodwin commented:
The only conflict was in the resource printer FWIW.

dgoodwin commented:
@deads2k can do, one of them sounds similar to something in Clayton's email. I'll get issues up for both though one is kinda specific to my PR, not sure it can be reproduced otherwise. I'll start with the one that can though.

dgoodwin commented:
The main one is reported in: https://github.com/openshift/origin/issues/7085

I'll try to work a bit on the one that *might* be a problem with my PR and see if I can tell what it is or explain in a relevant way for an issue.

deads2k commented:
@pweil- I blocked this on the rebase last time.  If its still good, he'd probably like a merge before the next one goes in :)

pweil- commented:
I will give this one more review tomorrow and push it before I go any further on the next rebase.

pweil- commented:
[merge]
